### PR TITLE
Adds redaction functions & rule definitions

### PR DESCRIPTION
**Pull request by wgehring-ml**
I'm not sure if I'm redacting all the right data (based rule definitions off of sled-envision-mes), so if you see something that should/should not be redacted let me know.

### CODE REVIEW

**Comment by wgehring-ml** [comment]
Already caught some identifiers that weren't being redacted in patients, will add a commit to redact those once I have checked for other missing redactions

File: src/main/ml-modules/root/lib/redaction/redactionFunctions.sjs [hunk 1]
```diff
@@ -0,0 +1,285 @@
...
+    xdmp.invokeFunction(
+      () => fn.head(
+        cts.valueMatch(
+          cts.jsonPropertyReference('redactMapEntries'),
+          `${sourceValue}:*`,
+          null,
+          cts.collectionQuery(mapCollection),
+        ),
+      ),
+      { database: xdmp.database('%%mlContentDatabaseName%%'), update: 'false' },
```
Comment (damonfeldman) [inline src/main/ml-modules/root/lib/redaction/redactionFunctions.sjs:44]: Usually we don't have gradle substitutions in the code itself. Won't redaction and the redacted values all be in the Final DB?

File: src/main/ml-modules/root/lib/redaction/redactionFunctions.sjs [hunk 2]
```diff
@@ -0,0 +1,285 @@
...
+
+/**
+ * Retrieve a cached value
+ *
+ * @param  {string}    name        The name of the cached value to retrieve
+ * @param  {Function}  getDefault  A function to provide a default value if there is not a cached value already
+ *
+ * @return {unknown}
+ */
+function getCachedValue(name, getDefault) {
```
Comment (damonfeldman) [inline src/main/ml-modules/root/lib/redaction/redactionFunctions.sjs]: Suggest we rename getDefault to findValueFn or similar. A default suggests that there is one value to use when there is a cache miss, but this function will probably be different for every key/name.

File: src/main/ml-modules/root/lib/redaction/redactionFunctions.sjs [hunk 3]
```diff
@@ -0,0 +1,285 @@
...
+      ),
+      { database: xdmp.database('%%mlContentDatabaseName%%'), update: 'false' },
+    ),
+  );
+}
+
+/**
+ * Retrieve a cached value
+ *
+ * @param  {string}    name        The name of the cached value to retrieve
```
Comment (damonfeldman) [inline src/main/ml-modules/root/lib/redaction/redactionFunctions.sjs]: Suggest we use key to emphasize this is basically a map.

File: src/main/ml-modules/root/lib/redaction/redactionUtils.sjs [hunk 4]
```diff
@@ -0,0 +1,487 @@
+const LOCK = Symbol('lock');
```
Comment (damonfeldman) [inline src/main/ml-modules/root/lib/redaction/redactionUtils.sjs:1]: Looks like solid code, but the reader (or me at least) has a hard time figuring out how the ciphers work. At the end of the day a couple key things should be more clear

* Every type of cypher has a set of chars that it applies to. Other chars are left as-is to preserve formatting.
* The basic cipher algorithm is to choose an array of offsets and use the same offset for the Nth character in the input string. Where an "offset" is the number of characters or digits to advance the input char (not sure if this is numeric, or offset into the allowed chars array)

Comments may clear a lot up, but fewer levels of indirection may also make it easier to use and extend the library.

Some distinction between deterministic and non-colliding should be made. The ciphers are all non-colliding, I think. If collisions are ok, a more basic random value will be more secure, at the cost of collisions.

**Pull request merged**